### PR TITLE
Polars: add strptime

### LINKED ISSUE #1755: Polars: year, month, timezone, offsets, etc
Polars: year, month, timezone, offsets, etc
**Description**
- **Constructor**: Transformation
- **Class**: Expr
- **DSL Variant:** Function: TemporalExpr
--- 
reference: [Polars APIs sheet](https://docs.google.com/spreadsheets/d/1w4H-qrSe9pkvNT2NhowfiIHNWXSseOyMX2nu5H6-RkQ/edit?gid=0#gid=0)

### PR DESCRIPTION
**Pull request by Shoeboxam**
Towards #1755

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/2096?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2096** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2096?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#2095** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2095?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2091** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2091?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2089** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2089?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: rust/src/domains/polars/datetime/mod.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+use polars::prelude::TimeUnit;
+
+use crate::{core::Domain, error::Fallible};
+
+/// A domain that represents a datetime.
+#[derive(Debug, Clone, PartialEq)]
+pub struct DatetimeDomain {
+    pub time_unit: TimeUnit,
+    pub time_zone: Option<String>,
```
Comment (mccalluc) [inline rust/src/domains/polars/datetime/mod.rs:9]: This seems odd, but I'm not sure I understand the goal. Expected a composition of Date and Time. Are we expecting to get anything out of the timezone string? Should there be hours ahead or behind UTC?

File: rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    ) {
+        return fallible!(
+            MakeTransformation,
+            "str.strptime dtype must be Time, Datetime or Date, found {}",
+            to_type
+        );
+    }
+    // Strict casting makes the transformation unstable: errors tell you things about private data.
+    // Could throw an error if strict, but it is the default, so for ease-of-use it is forced to be non-strict.
+    // It is also ok for overflow to wraparound.
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs:61]: What is the range of dates which we can represent? If dates earlier than 1970 wrap around, that is going to be undesirable in many cases.

File: rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            "non-exact not implemented for Time data type"
+        );
+    }
+
+    // never raise on error
+    let ambiguous_str = literal_value_of(ambiguous)
+        .ok()
+        .flatten()
+        .unwrap_or("raise".to_string());
+    let ambiguous = if ["earliest".to_string(), "latest".to_string()].contains(&ambiguous_str) {
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs:80]: "earliest" and "latest" don't occur elsewhere in this PR. Are they coming from a library?

File: rust/src/transformations/make_stable_expr/expr_str/expr_strptime/test.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                DataType::Date,
+                StrptimeOptions {
+                    format: Some("%Y-%m-%d".to_string()),
+                    strict: false,
+                    exact: false,
+                    cache: true,
+                },
+                lit("null"),
+            )
+            .alias("date"),
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_str/expr_strptime/test.rs:69]: Do you need the alias if the name doesn't change?

File: rust/src/transformations/make_stable_expr/expr_str/expr_strptime/test.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+┌──────────┬─────────────────────┬────────────┐
+│ time     ┆ datetime            ┆ date       │
+│ ---      ┆ ---                 ┆ ---        │
+│ time     ┆ datetime[ns]        ┆ date       │
+╞══════════╪═════════════════════╪════════════╡
+│ 14:00:00 ┆ 2023-08-12 14:00:00 ┆ 2023-08-12 │
+│ 15:00:00 ┆ 2024-09-13 15:00:00 ┆ 2024-09-13 │
+│ 16:00:00 ┆ 2025-10-14 16:00:00 ┆ 2025-10-14 │
+│ 16:00:00 ┆ null                ┆ null       │
+└──────────┴─────────────────────┴────────────┘"#;
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_str/expr_strptime/test.rs:84]: This kind of test is going to be fragile if the data changes... but it might be easier than constructing a dataframe to check against.

File: rust/src/transformations/make_stable_expr/expr_str/expr_strptime/test.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            "2023-08-12 14:00:00".to_string(),
+            "2024-09-13 15:00:00".to_string(),
+            "2025-10-14 16:00:00".to_string(),
+            "2025-13-14 16:00:00".to_string()
+        ],
+        "date" => &[
+            "2023-08-12".to_string(),
+            "2024-09-13".to_string(),
+            "2025-10-14".to_string(),
+            "2025-13-14".to_string()
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_str/expr_strptime/test.rs:31]: Should you test behavior with bad data? Ill-formated, day-of-month out of range, overflow, underflow.

**Review by mccalluc — COMMENTED** [review]
Main concern is how ill-formed / out-of-range dates get handled.

File: rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs (hunk 3 above)
Comment (mccalluc) [reply rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs:80]: Never mind: found https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.str.strptime.html

File: rust/src/domains/polars/datetime/mod.rs (hunk 1 above)
Comment (Shoeboxam) [reply rust/src/domains/polars/datetime/mod.rs:9]: * Time (i64): number of nanoseconds since midnight
* Datetime (i64): number of milli/nano/micro seconds since Unix epoch. Time zone and time unit are encoded as part of the type/shared amongst all rows in a column
* Date (i32): number of days since Unix epoch.

Time and Date don't carry any extra type information/domain descriptors, so they can be represented via AtomDomain.
Datetime is separate as it isn't a composition of date and time, it is its own data format. It is also different in that it has type information (zone and unit) that needs to be known to interpret the meaning of the data. For the purposes of OpenDP, I consider the set of time points spanned by the domain to be distinct for each time zone. Updated the documentation.

File: rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs (hunk 2 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs]: Dates and datetimes are signed offsets from the epoch, so dates earlier than 1970 don't wrap until much earlier time points.

**Comment by Shoeboxam** [comment]
I got rid of the fixtures, rebased on top of the refactored ExprDomain, and updated the tests to edge cases per your suggestions @mccalluc. Thanks for the thoughtful review!

File: rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/mod.rs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+    if !matches!(
+        to_type,
+        DataType::Time
+            | DataType::Datetime(TimeUnit::Microseconds | TimeUnit::Milliseconds, _)
+            | DataType::Date
+    ) {
+        return fallible!(
+            MakeTransformation,
+            "str.strptime output dtype must be Time, micro- or milli-second Datetime or Date, found {}",
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/mod.rs:96]: The comment above also mentions nanoseconds. Add to this message?

File: rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/test.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        StrptimeOptions {
+            format: Some("%Y-%m-%dT%H:%M:%SZ".to_string()),
+            strict: false,
+            exact: true,
+            cache: true,
+        },
+        lit("null"),
+    ));
+
+    println!("{:?}", lf.collect());
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/test.rs:94]: Stricter test here?

File: rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/test.rs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    assert_eq!(format!("{:?}", observed), expected);
+    Ok(())
+}
+
+#[test]
+fn test_strptime_datetime() -> Fallible<()> {
+    let lf = df!("datetime" => &[
+        "+262142-01-01T00:00:00Z",
+        "+262143-01-01T00:00:00Z",
+        "+262144-01-01T00:00:00Z",
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/test.rs:80]: These are unusual years: Any reason to pick these particular three? I feel like it's trying to exercise something particular, but I'm not sure what.

File: rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/mod.rs (hunk 7 above)
Comment (mccalluc) [reply rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/mod.rs:96]: Oh, the absence of nano is probably related to https://github.com/pola-rs/polars/issues/19928. Maybe file an issue in this repo to track that?

File: rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/test.rs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+                format: Some("%Y-%m-%dT%H:%M:%S.fZ".to_string()),
+                strict: false,
+                exact: true,
+                cache: true,
+            },
+            lit("null"),
+        ));
+
+    let expected = DataFrame::new(vec![Series::new("t", vec![None::<NaiveTime>])])?;
+    assert_eq!(lf.collect()?, expected);
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/test.rs:115]: If I'm reading this correctly, there's no data in the expected frame. Is this because there's a problem with the decimal seconds? This behavior is really different from the integer seconds that follow. Consider splitting this into two tests, and either in the test name or a comment, explain why this one is different.

File: rust/src/transformations/make_stable_expr/namespace_str/mod.rs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    match str_function {
+        #[cfg(feature = "contrib")]
+        StringFunction::Strptime(_, _) => expr_strptime::make_expr_strptime(input_domain, input_metric, expr),
+
+        expr => fallible!(
+            MakeTransformation,
+            "Expr is not recognized at this time: {:?}. {}If you would like to see this supported, please file an issue.",
+            expr,
+            get_disabled_features_message()
+        )
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_str/mod.rs:50]: This error seems to be conflating two separate things. Maybe they just don't have `contrib` enabled? Or maybe even with `contrib`, it doesn't match?

Maybe have two separate `fallible!`s, one guarded with `#[cfg(feature = "contrib")]`, and one with the negation?

Or if it's better to keep it simple, one error message that makes it clear that there are two different possible root causes?

**Review by mccalluc — COMMENTED** [review]
Thanks for the explanations. Reviewed from the top, and there are still some things that seem odd to me, but I might be confused.

Thinking about scope: I'm not sure if this would be exposed on the python side with this PR, or if there is follow-up work that will do that. If there isn't a follow-up, then it would be nice to have some tests on the Python side, even if we're holding off on docs until the feature set for this release is stabilized.

File: rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/mod.rs (hunk 7 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/mod.rs:96]: I left an error condition in the code.

File: rust/src/transformations/make_stable_expr/namespace_str/mod.rs (hunk 11 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/namespace_str/mod.rs:50]: This is what the `get_disabled_features_message()` call does. If there are disabled features it interpolates that explanation into the error message.

**Review by mccalluc — APPROVED** [review]
Just looking at the latest changes, and the explanations are very helpful!
